src/mds/MDCache.cc

```diff
@@ -10204,7 +10204,10 @@ struct C_MDC_RetryScanStray : public MDCacheContext {
   }
 };
 
-void MDCache::scan_stray_dir(dirfrag_t next)
+// if the formatter is not null, the request comes from asok and formatter will
+// accumulate stray dirs info
+// the function returns EAGAIN to support waiting when called from asok
+int MDCache::scan_stray_dir(dirfrag_t next, Formatter* f)
```

Comment:
I think you need to adjust this to complete a context when complete so that you can call `on_finish` for the asok command completion.